packages/protocol/src/interaction/InteractionClient.ts

```diff
@@ -492,61 +480,66 @@ export class InteractionClient {
+            `Sending write request: ${attributes
+                .map(
+                    ({ endpointId, clusterId, attribute: { id }, value, dataVersion }) =>
+                        `${resolveAttributeName({ endpointId, clusterId, attributeId: id })} = ${Logger.toJSON(
+                            value,
+                        )} (version=${dataVersion})`,
+                )
+                .join(", ")}`,
+        );
+        const writeRequests = attributes.map(
+            ({ endpointId, clusterId, attribute: { id, schema }, value, dataVersion }) => ({
+                path: { endpointId, clusterId, attributeId: id },
+                data: schema.encodeTlv(value, { forWriteInteraction: true }),
+                dataVersion,
+            }),
+        );
+        const timedRequest =
+            attributes.some(({ attribute: { timed } }) => timed) ||
+            asTimedRequest === true ||
+            options.timedRequestTimeoutMs !== undefined;
+
+        const response = await this.withMessenger<TypeFromSchema<typeof TlvWriteResponse> | undefined>(
+            async messenger => {
+                if (timedRequest) {
+                    await messenger.sendTimedRequest(timedRequestTimeoutMs);
                 }
-                return [];
+
+                return await messenger.sendWriteCommand({
+                    suppressResponse,
+                    timedRequest,
+                    writeRequests,
+                    moreChunkedMessages: false,
+                    interactionModelRevision: Specification.INTERACTION_MODEL_REVISION,
+                });
+            },
+            executeQueued,
+        );
+
+        if (response === undefined) {
+            if (!suppressResponse) {
+                throw new MatterFlowError(`No response received from write interaction but expected.`);
             }
-            return response.writeResponses
-                .flatMap(
-                    ({ status: { status, clusterStatus }, path: { nodeId, endpointId, clusterId, attributeId } }) => {
-                        return {
-                            path: { nodeId, endpointId, clusterId, attributeId },
-                            status: status ?? clusterStatus ?? StatusCode.Failure,
-                        };
-                    },
-                )
-                .filter(({ status }) => status !== StatusCode.Success);
-        }, executeQueued);
+            return [];
+        }
+        return response.writeResponses
+            .flatMap(({ status: { status, clusterStatus }, path: { nodeId, endpointId, clusterId, attributeId } }) => {
+                return {
+                    path: { nodeId, endpointId, clusterId, attributeId },
+                    status: status ?? clusterStatus ?? StatusCode.Failure,
+                };
+            })
+            .filter(({ status }) => status !== StatusCode.Success);
     }
 
     async subscribeAttribute<A extends Attribute<any, any>>(options: {
```

Comment:
AsyncIterator probably makes sense for all of these subscribe functions... Then internally you only have to resolve the next() promise when a report comes in and the callbacks are unnecessary... Maybe something to consider when we resolve the interactionclient/interactionserver thing